### ISSUE TITLE
Can not select negated boolean column in MS SQL Server

### ISSUE THREAD

**Issue by mkapi** [sqlalchemy/sqlalchemy#5352]
I would like negate a column of `BIT` type and use the resulting column in a select query. To do this I use the invert `~` operator from sqlalchemy.

The code below demonstrates this bug. The second statement is compiled into `SELECT a.flag = 0 AS anon_1 FROM a` which can not execute and throws the error `SQL Error [102] [S0001]: Incorrect syntax near '='.`.
According to MS SQL Server docs the [Bitwise NOT (~) operator](https://docs.microsoft.com/en-us/sql/t-sql/language-elements/bitwise-not-transact-sql) can negate `BIT` columns and the not `~` operator is supported in SQLAlchemy. So the statement `SELECT ~a.flag AS anon_1 FROM a` works and gives correct results.

I've come across #4250 which seems related. 

```py
from sqlalchemy import *
from sqlalchemy.orm import *
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()


class A(Base):
    __tablename__ = 'a'
    id = Column(Integer, primary_key=True)
    flag = Column(Boolean)


engine = create_engine('mssql+pymssql://user:pass@host:port/db', echo=True)
Session = sessionmaker(engine)

Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

conn = engine.connect()

engine.execute(select([A.flag])).fetchall()
engine.execute(select([~A.flag])).fetchall()

```

**Versions.**
 - OS: Ubuntu 18.04.4 LTS
 - Python: Python 3.7.5
 - SQLAlchemy: 1.3.12
 - Database: Microsoft SQL Server 2017 14.0.3294.2
 - DBAPI: pymssql

**Have a nice day!**

**Comment by zzzeek** [comment]
hi there -

how interesting that the exact test case from #4250 is still failing though I would assume in a different way from what was fixed there.

this limitation has more to do with SQL Server not being able to SELECT from an expression rather than the syntax being invalid.   SQLAlchemy has a single style for treatment of "non-native booleans" which is to compare them to 1 or 0, and to add a third methodology here might be very involved.

The syntax is valid in the WHERE clause:

SELECT a.flag 
FROM a 
WHERE a.flag = 0

SQL Server just can't SELECT from expressions:

```
conn.execute("SELECT 1 > 2").fetchall()
SELECT 1 > 2

Traceback (most recent call last):
pyodbc.ProgrammingError: ('42000', "[42000] [Microsoft][ODBC Driver 13 for SQL Server][SQL Server]Incorrect syntax near '>'. (102) (SQLExecDirectW)")


```

therefore I'm not sure how much use one would get out of adding this very specific exception case.  While we could with effort make this exact use case of a negated boolean expression work in the columns clause, we would still not be helping the general case of allowing "expressions" to work in the columns clause.    An approach that allows arbitrary expressions to work that would be applied to SQL Server would be much more worthwhile.

to get your exact expression we can use a custom op which is somewhat inaccurately described at https://docs.sqlalchemy.org/en/13/core/sqlelement.html?highlight=custom_op#sqlalchemy.sql.operators.custom_op (imports are wrong):

```
from sqlalchemy.sql.elements import UnaryExpression
from sqlalchemy.sql import operators

def bit_negate(expr):
    return UnaryExpression(expr,
        operator=operators.custom_op("~"),
        type_=Boolean)



conn.execute(select([bit_negate(A.flag)]).where(~A.flag)).fetchall()


```

**Comment by mkapi** [comment]
Thanks for the workaround. It is unclear to me why SQL Server can't translate a boolean expression into BIT type which seems straightforward.

I guess I should close this issue, no?

**Comment by zzzeek** [comment]
maybe I can look to see how hard this is to do.    we can leave open for awhile.

**Comment by CaselIT** [comment]
is there a way for the compiler to know if it is rendering a where/having/join clause or something else?

**Comment by zzzeek** [comment]
it can know that sure but maybe in this case we add compilation rule to mssql that compiles neg_op as "~<expr>".   so it would change everywhere.

**Comment by CaselIT** [comment]
yes, that seems more appropriate. it seems that mssql has no concept of boolean, and there is not way of casting to boolean. That seems also the reason that `select 1>2` does not work, while `select 1+ 2` does. It has no way or representing a boolean.
It does not like `~ a.flag` in a where clause, because it has type bit (or int, or whatever `flag` type has) and it needs boolean in the where.

**Comment by zzzeek** [comment]
ohhh ~flag doesnt work in the WHERE clause?   well yeah then we have to use the **kwargs passed around the compiler to know where we are, that's inconvenient.   might have to poke around to see if we have a consistent flag that will tell us this part including for subqueries.

**Comment by CaselIT** [comment]
`~flag` by itself it does not work since it's of type bit (or int etc) and not boolean.
Using `~flag = 1` does work.

**Comment by zzzeek** [comment]
but it still has to be different in the columns clause vs the WHERE clause, right?  that's the whole thing here.

**Comment by CaselIT** [comment]
yes, we do not want to use `select ~flag = 1` instead of `select ~flag` since that would return a different result (other than the fact that mssql errors out)

**Comment by zzzeek** [comment]
there's a kwarg within_columns_clause=True that will be in the compilation methods for this.

so here it is:

```
diff --git a/lib/sqlalchemy/dialects/mssql/base.py b/lib/sqlalchemy/dialects/mssql/base.py
index 5e0704597..3052ec933 100644
--- a/lib/sqlalchemy/dialects/mssql/base.py
+++ b/lib/sqlalchemy/dialects/mssql/base.py
@@ -1598,6 +1598,14 @@ class MSSQLCompiler(compiler.SQLCompiler):
 
         return decorate
 
+    def visit_isfalse_unary_operator(self, element, operator, **kw):
+        if kw.get("within_columns_clause", False):
+            return "~%s" % self.process(element.element, **kw)
+        else:
+            return super(MSSQLCompiler, self).visit_isfalse_unary_operator(
+                element, operator, **kw
+            )
+
     def visit_now_func(self, fn, **kw):
         return "CURRENT_TIMESTAMP"
 

```

poc
```

from sqlalchemy import Boolean
from sqlalchemy import column
from sqlalchemy import select
from sqlalchemy.dialects import mssql

c = column('x', Boolean)

stmt = select([~c]).where(~c)
print(stmt.compile(dialect=mssql.dialect()))

```

**Comment by zzzeek** [comment]
```
SELECT ~x AS x 
WHERE x = 0
```

**Comment by zzzeek** [comment]
however the above patch needs to assert that the type of the expression is Boolean.  otherwise it should just pass through so that SQL Server can sort it out.

**Comment by CaselIT** [comment]
Sorry I may have misunderstood something.
Do we want `~` to render as a bitwise not or as a `not`?
Because `select ~x` will render a bitwise not and that works with any int type

https://docs.microsoft.com/en-us/sql/t-sql/language-elements/bitwise-not-transact-sql?view=sql-server-ver15
